File: game-engine/src/main/kotlin/com/failureludo/engine/Board.kt
package com.failureludo.engine

/**
 * Immutable board constants.
 *
 * The main track has 52 cells (indices 0-51) shared by all players.
 * Each player enters at their own ENTRY_POSITIONS index and travels clockwise.
 * The HOME_COLUMN_ENTRY is the last main-track index a piece visits before
 * branching into the colour-specific home column (5 steps → Finished).
 *
 * Safe squares: pieces on these cells CANNOT be captured.
 *   • All four entry squares (starting squares of each colour)
 *   • The four "starred" squares just before each home-column entrance
 */
object Board {

    const val MAIN_TRACK_SIZE = 52
    const val HOME_COLUMN_STEPS = 5  // steps 1-5; step 6 would be Finished

    /** Main-track index where each colour enters the board. */
    val ENTRY_POSITIONS: Map<PlayerColor, Int> = mapOf(
        PlayerColor.RED    to 0,
        PlayerColor.BLUE   to 13,
        PlayerColor.YELLOW to 26,
        PlayerColor.GREEN  to 39
    )

    /**
     * Main-track index that is the LAST cell before the home column.
     * A piece on this cell with a die value of 1 moves to HomeColumn(1);
     * a die value of 2 → HomeColumn(2), etc. — if it fits within 5 steps.
     */
    val HOME_COLUMN_ENTRY: Map<PlayerColor, Int> = mapOf(
        PlayerColor.RED    to 50,
        PlayerColor.BLUE   to 11,
        PlayerColor.YELLOW to 24,
        PlayerColor.GREEN  to 37
    )

    /**
     * Safe squares on the main track (0-51).
     * Pieces here cannot be captured by opponents.
     */
    val SAFE_SQUARES: Set<Int> = setOf(
        0, 8,   // RED   entry + starred
        13, 21, // BLUE  entry + starred
        26, 34, // YELLOW entry + starred
        39, 47  // GREEN entry + starred
    )

    /**
     * Given a [color]'s [relative] travel distance from entry (0 = entry square),
     * returns the absolute main-track index.
     */
    fun absolutePosition(color: PlayerColor, relative: Int): Int =
        (ENTRY_POSITIONS.getValue(color) + relative) % MAIN_TRACK_SIZE

    /**
     * Given a piece's current absolute [mainTrackIndex] and its [color],
     * returns how many steps it has travelled from its entry point.
     */
    fun relativePosition(color: PlayerColor, mainTrackIndex: Int): Int {
        val entry = ENTRY_POSITIONS.getValue(color)
        return (mainTrackIndex - entry + MAIN_TRACK_SIZE) % MAIN_TRACK_SIZE
    }

    /**
     * Returns how many more main-track steps remain before a piece of [color]
     * reaches the home-column entry (i.e., how many steps until the branch point).
     * A piece exactly on HOME_COLUMN_ENTRY has 0 remaining main-track steps.
     */
    fun stepsToHomeColumnEntry(color: PlayerColor, currentAbsoluteIndex: Int): Int {
        val entryRelative = 0
        val maxRelative = relativePosition(color, HOME_COLUMN_ENTRY.getValue(color))
        val currentRelative = relativePosition(color, currentAbsoluteIndex)
        return (maxRelative - currentRelative + MAIN_TRACK_SIZE) % MAIN_TRACK_SIZE
    }

    fun isSafeSquare(index: Int): Boolean = index in SAFE_SQUARES
}

File: ludo-web/src/engine/board.ts
import type { PlayerColor } from './types';

export const MAIN_TRACK_SIZE = 52;
export const HOME_COLUMN_STEPS = 5;

export const ENTRY_POSITIONS: Record<PlayerColor, number> = {
  RED: 0,
  BLUE: 13,
  YELLOW: 26,
  GREEN: 39,
};

export const HOME_COLUMN_ENTRY: Record<PlayerColor, number> = {
  RED: 50,
  BLUE: 11,
  YELLOW: 24,
  GREEN: 37,
};

export const SAFE_SQUARES = new Set([0, 8, 13, 21, 26, 34, 39, 47]);

export function isSafeSquare(index: number): boolean {
  return SAFE_SQUARES.has(index);
}

export function relativePosition(color: PlayerColor, mainTrackIndex: number): number {
  const entry = ENTRY_POSITIONS[color];
  return (mainTrackIndex - entry + MAIN_TRACK_SIZE) % MAIN_TRACK_SIZE;
}

export function stepsToHomeColumnEntry(color: PlayerColor, currentAbsoluteIndex: number): number {
  const maxRelative = relativePosition(color, HOME_COLUMN_ENTRY[color]);
  const currentRelative = relativePosition(color, currentAbsoluteIndex);
  return (maxRelative - currentRelative + MAIN_TRACK_SIZE) % MAIN_TRACK_SIZE;
}

// Visual grid coordinates (15×15)
export const MAIN_TRACK_CELLS: [number, number][] = [
  // Red side (0-4): row 6, cols 1-5
  [6, 1], [6, 2], [6, 3], [6, 4], [6, 5],
  // Up col 6 (5-10): rows 5-0
  [5, 6], [4, 6], [3, 6], [2, 6], [1, 6], [0, 6],
  // Top row (11-12): row 0, cols 7-8
  [0, 7], [0, 8],
  // Down col 8 (13-17): rows 1-5
  [1, 8], [2, 8], [3, 8], [4, 8], [5, 8],
  // Blue side (18-23): row 6, cols 9-14
  [6, 9], [6, 10], [6, 11], [6, 12], [6, 13], [6, 14],
  // Right col 14 (24-25): rows 7-8
  [7, 14], [8, 14],
  // Yellow side (26-30): row 8, cols 13-9
  [8, 13], [8, 12], [8, 11], [8, 10], [8, 9],
  // Down col 8 (31-36): rows 9-14
  [9, 8], [10, 8], [11, 8], [12, 8], [13, 8], [14, 8],
  // Bottom row (37-38): row 14, cols 7-6
  [14, 7], [14, 6],
  // Green side (39-43): rows 13-9, col 6
  [13, 6], [12, 6], [11, 6], [10, 6], [9, 6],
  // Left row 8 (44-49): row 8, cols 5-0
  [8, 5], [8, 4], [8, 3], [8, 2], [8, 1], [8, 0],
  // Left col 0 (50-51): rows 7-6
  [7, 0], [6, 0],
];

export const HOME_COLUMNS: Record<PlayerColor, [number, number][]> = {
  RED:    [[7, 1], [7, 2], [7, 3], [7, 4], [7, 5]],
  BLUE:   [[1, 7], [2, 7], [3, 7], [4, 7], [5, 7]],
  YELLOW: [[7, 13], [7, 12], [7, 11], [7, 10], [7, 9]],
  GREEN:  [[13, 7], [12, 7], [11, 7], [10, 7], [9, 7]],
};

export const HOME_YARD_SPOTS: Record<PlayerColor, [number, number][]> = {
  RED:    [[1, 1], [1, 4], [4, 1], [4, 4]],
  BLUE:   [[1, 10], [1, 13], [4, 10], [4, 13]],
  YELLOW: [[10, 10], [10, 13], [13, 10], [13, 13]],
  GREEN:  [[10, 1], [10, 4], [13, 1], [13, 4]],
};

export const CENTER: [number, number] = [7, 7];

export function colorIndex(color: PlayerColor): number {
  return { RED: 0, BLUE: 1, YELLOW: 2, GREEN: 3 }[color];
}

export function teamIndex(color: PlayerColor): number {
  return color === 'RED' || color === 'YELLOW' ? 0 : 1;
}
